Convert /domain/navigation to typescript

MidhunSureshR opened the pull request.

MidhunSureshR committed:
WIP

bwindels commented on src/domain/navigation/Navigation.ts:
```diff
@@ -15,28 +15,53 @@ limitations under the License.
 */
 
 import {BaseObservableValue, ObservableValue} from "../../observable/ObservableValue";
+import type {allowsChild as AllowsChild} from "./index.js";
```
I'd write the signature of the allowsChild in here as a type. Best not to have the definition be the implementation. You can probably also define it so it just takes keys of (a generic) SegmentType.

bwindels commented on src/domain/navigation/Navigation.ts:
```diff
@@ -60,7 +85,7 @@ export class Navigation {
     }
 
-    observe(type) {
+    observe(type: keyof SegmentType): SegmentObservable {
```
Probably best to make Navigation generic over the SegmentType, so it can be used for different setups (e.g. in apps built on the sdk). EventEmitter is an example of how to do this. The allowsChild function would also be generic over the segment type, so that way you can enforce that the allowsChild function is for the right segment type.

MidhunSureshR committed:
WIP - 2

MidhunSureshR committed:
Complete converting Navigation.js to ts

MidhunSureshR committed:
Type allowsChild

MidhunSureshR committed:
Convert index.js to typescript

MidhunSureshR committed:
Remove deprecated method

MidhunSureshR committed:
Make all keys string

MidhunSureshR committed:
Fix formatting

MidhunSureshR committed:
Type function

MidhunSureshR committed:
Convert URLRouter.js to typescript

MidhunSureshR committed:
Pass as separate arguments to constructor

MidhunSureshR committed:
Use segment type

MidhunSureshR committed:
Type createRouter function

MidhunSureshR committed:
Make defaultSessionId optional

MidhunSureshR committed:
Extract complex type as type alias

MidhunSureshR committed:
Mark methods as private

MidhunSureshR committed:
Use subtype instead of whole SegmentType

MidhunSureshR committed:
Use undefined instead of null

MidhunSureshR committed:
Remove comment

MidhunSureshR committed:
Remove empty line

MidhunSureshR committed:
Add return type

MidhunSureshR committed:
Remove unwanted export

MidhunSureshR committed:
WIP

MidhunSureshR committed:
Give defaultt type

bwindels commented on src/domain/navigation/Navigation.ts:
```diff
@@ -16,27 +16,39 @@ limitations under the License.
+
+export type OptionalValue<T> = T extends true? [(undefined | true)?]: [T];
+
+export class Navigation<T> {
```
if you use `keyof T`, don't you need to say T extends object?

bwindels commented on src/domain/ViewModel.ts:
```diff
@@ -27,17 +27,18 @@ import type {Platform} from "../platform/web/Platform";
 }
 
-export class ViewModel<O extends Options = Options> extends EventEmitter<{change: never}> {
+export class ViewModel<N extends {session: string} = SegmentType, O extends Options<N> = Options<N>> extends EventEmitter<{change: never}> {
```
where is `session` needed?

bwindels replied:
Ah, because of `_getLastSessionId` in URLRouter. Can we add this constraint in URLRouter rather than here? That way, we aren't constrained if we don't use the URLRouter.

MidhunSureshR replied:
Typescript complains if we don't duplicate the constraint here. I've made a [change](https://github.com/vector-im/hydrogen-web/pull/676/commits/48fd2c3aa25eee1f2328e5febf8feddeda229821) that would allow other projects to use ViewModel even if they use a SegmentType that lacks support for session segment.

bwindels commented on src/domain/ViewModel.ts:
```diff
@@ -131,11 +132,11 @@ export class ViewModel<O extends Options = Options> extends EventEmitter<{change
     }
 
-    get navigation(): Navigation {
+    get navigation(): Navigation<N> {
```
remind me, the reason we use a generic type here is so that in projects that use the sdk, they can use their own segment type?

MidhunSureshR replied:
Exactly. The navigation is mostly always going to be accessed via this getter. One downside of this is that all viewmodels in projects that use hydrogen SDK will always need to specify the type parameter.

bwindels commented on src/domain/navigation/Navigation.ts:
```diff
@@ -16,27 +16,39 @@ limitations under the License.
+
+type AllowsChild<T> = (parent: Segment<T> | undefined, child: Segment<T>) => boolean;
+
+export type OptionalValue<T> = T extends true? [(undefined | true)?]: [T];
```
👀 this one is taking some effort to decipher... what do the square brackets do? It's not an array if I'm not mistaken.

bwindels replied:
It's a [tuple](https://www.typescriptlang.org/docs/handbook/2/objects.html#tuple-types), can you elaborate a bit why this is needed/how it is used?

MidhunSureshR replied:
Yeah this is a bit unsightly.
It is used like this (code from Segment ctor):
```js
    constructor(public type: K, ...value: OptionalValue<T[K]>) {
        this.value = (value[0] === undefined ? true : value[0]) as unknown as T[K];
    }
```
The `value` parameter is made a [rest parameter](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html#generic-parameter-defaults) so that we can make it both optional and give it a default value of undefined depending on the segment type.
So for example, these should be accepted:
```js
new Segment("create-room"); // value is optional, internally given the value true
new Segment("create-room", true);
```
where as these should not be accepted:
```js
new Segment("room") // error: second argument is needed and it should be string (room-id)
```
`OptionalValue` is basically stating that if `SegmentType[type] = true`:
1. Allow this argument to be optional
2. Give it a default value of undefined
3. Also allow it to be true

I tried a lot of things but they would all require that we explicitly pass undefined as second argument, eg:
```js
constructor(public type: K, value: T[K] extends true? undefined: T[K]) {
    this.value = (value[0] === undefined ? true : value[0]) as unknown as T[K];
}

const s: Segment<SegmentType> = new Segment("create-room"); // error: An argument for 'value' was not provided
const s: Segment<SegmentType> = new Segment("create-room", undefined); //works, but not what we want
```
or they would make the second parameter always optional.

bwindels replied:
right, makes sense! Can you perhaps put a summary of this in a comment?

bwindels reviewed (commented):
Had a first look and looks great. Just a few questions.

MidhunSureshR committed:
Generic parameter should extend object

MidhunSureshR committed:
Make URLRouter in options conditional on generic

URLRouter can be passed in option to vm only if the SegmentType used
contains session.
ViewModel.urlCreator returns undefined when used with a SegmentType that
lacks session.

MidhunSureshR committed:
WIP

MidhunSureshR committed:
Fix type in observeNavigation

bwindels commented on src/domain/LogoutViewModel.ts:
```diff
@@ -16,10 +16,11 @@ limitations under the License.
 type LogoutOptions = { sessionId: string; } & Options;
 
-export class LogoutViewModel extends ViewModel<LogoutOptions> {
+export class LogoutViewModel extends ViewModel<SegmentType, LogoutOptions> {
```
Would it make sense to swap the order of the generic params in ViewModel? I think within Hydrogen, once everything is TS-ified, every view model will specify the options type, but I _think_ none will explicitly set the segment type? IIRC, that would avoid have to the above in most view models (once TS).

MidhunSureshR replied:
```ts
export class ViewModel<N extends object = SegmentType, O extends Options<N> = Options<N>> extends EventEmitter<{change: never}>
```
Not sure if it can be done that way since the Options generic is dependent on the first one.

bwindels reviewed (commented):
Apart from comments, looks good to me!

MidhunSureshR committed:
Add explaining comment

bwindels committed:
Merge branch 'master' into ts-conversion-domain-navigation
